src/lib/storage/lz4/lz4_encoder.hpp

```diff
@@ -122,48 +137,348 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
+    dictionary.shrink_to_fit();
+
+    return dictionary;
+  }
+
+  /**
+   * When training a dictionary for strings, each row element is a sample. This is not done when training a
+   * dictionary for non-strings since zstd's dictionary is made for strings (and the samples can have different sizes).
+   * First, a dictionary is trained with the provided data and sample sizes. If this does not succeed, we try to
+   * increase the input size by repeating the values and adding larger sample sizes up to a certain limit. If this still
+   * fails or the input size is too small in general, a dictionary won't be trained and can't be used for compression.
+   * Compressing the blocks independently and without a dictionary hurts the compression ratio.
+   *
+   * @param values The input data that will be compressed (i.e., all strings concatenated).
+   * @param sample_sizes A vector of sample lengths. Each length corresponds to a substring in the values vector. These
+   *                     should correspond to the length of each row's value.
+   * @return The trained dictionary or in the case of failure an empty vector.
+   */
+  pmr_vector<char> _train_string_dictionary(const pmr_vector<char>& values, const pmr_vector<size_t>& sample_sizes) {
+    /**
+     * The recommended dictionary size is about 1/100th of size of all samples combined, but he size also has to be at
+     * least 1KB. Smaller dictionaries won't work.
+     */
+    auto max_dictionary_size = values.size() / 100;
+    max_dictionary_size = std::max(max_dictionary_size, _minimum_dictionary_size);
+
+    auto dictionary = pmr_vector<char>{values.get_allocator()};
+    size_t dictionary_size;
+
+    // If the input does not contain enough values, it won't be possible to train a dictionary for it.
+    if (values.size() < _minimum_value_size) {
+      return dictionary;
+    }
+
+    dictionary.resize(max_dictionary_size);
+    dictionary_size = ZDICT_trainFromBuffer(dictionary.data(), max_dictionary_size, values.data(), sample_sizes.data(),
+                                            static_cast<unsigned>(sample_sizes.size()));
+
+    // If the generation failed, try generating a dictionary with more input.
+    if (ZDICT_isError(dictionary_size)) {
+      return _train_string_dictionary_padded(values, sample_sizes, max_dictionary_size);
+    }
+
+    DebugAssert(dictionary_size <= max_dictionary_size,
+                "Generated ZSTD dictionary in LZ4 compression is larger than "
+                "the memory allocated for it.");
+
+    // Shrink the allocated dictionary size to the actual size.
+    dictionary.resize(dictionary_size);
+    dictionary.shrink_to_fit();
+
+    return dictionary;
+  }
+
+  /**
+   * Increase the dictionary input data by appending the original input data (linear increase) and increasing the
+   * maximum dictionary size (exponential increase). This only happens if the input data is too small for zstd to
+   * successfully train a dictionary.
```

Comment:
Is this a standard approach or did you come up with it?

Reply:
It's our own workaround. We had some trouble with the dictionary training, as it fails when it doesn't find enough redundant information in the given data for the given target dictionary size, which it's documentation says it can just do. We found that increasing the size of the dictionary and/or duplicating the input helps with that, but it seams that there are still cases where it will just fail for reasons not really not known to us. We will discuss that at our next meeting on Tuesday, it might be wiser to not train a dictionary if it fails the first time and just take a loss in compression ratio that comes with not using one.
Just fyi, my teammates and me are all currently in different places on different missions this week and can't really work on implementing your reviews, but we will catch up on the weekend / beginning of next week.

Reply:
Since this workaround segfaults with a large amount of data and doesn't really work without the magic line, we removed it.